core/shared/src/main/scala/fs2/compression/Gzip.scala

```diff
@@ -0,0 +1,631 @@
+              Pull.output1(
+                (
+                  Option.empty[String],
+                  crc32Builder,
+                  Stream.raiseError(new ZipException(s"Failed to read $fieldName field"))
+                )
+              )
+          }
+          .stream
+      else Stream.emit((Option.empty[String], crc32Builder, stream))
+
+  private def _gunzip_validateHeader(
+      isPresent: Boolean,
+      headerCrc32: CrcBuilder[BitVector]
+  ): Pipe[F, Byte, Byte] =
+    stream =>
+      if (isPresent)
+        stream.pull
+          .unconsN(gzipHeaderCrcBytes)
+          .flatMap {
+            case Some((headerCrcChunk, streamAfterHeaderCrc)) =>
+              val expectedHeaderCrc16 = unsignedToInt(headerCrcChunk(0), headerCrcChunk(1))
+              val actualHeaderCrc16 = headerCrc32.result.toInt() & 0xffff
+              if (expectedHeaderCrc16 != actualHeaderCrc16)
+                Pull.raiseError(new ZipException("Header failed CRC validation"))
+              else
+                Pull.output1(streamAfterHeaderCrc)
+            case None =>
+              Pull.raiseError(new ZipException("Failed to read header CRC"))
+          }
+          .stream
+          .flatten
+      else stream
+
+  private def _gunzip_validateTrailer(
+      trailerStream: Ref[F, Chunk[Byte]],
+      crc32: Ref[F, Long],
+      actualInputSize: Ref[F, Long]
+  ): Stream[F, Byte] =
+    Stream
+      .eval {
+        trailerStream.get.flatMap[Unit] { trailerChunk =>
+          if (trailerChunk.size != gzipTrailerBytes) {
+            F.raiseError(new ZipException(s"Failed to read trailer (1): $trailerChunk"))
+          } else {
+            crc32.get.flatMap { crc32 =>
+              actualInputSize.get.flatMap { actualInputSize =>
+                val expectedInputCrc32 =
+                  unsignedToLong(
+                    trailerChunk(0),
+                    trailerChunk(1),
+                    trailerChunk(2),
+                    trailerChunk(3)
+                  ) & 0xffffffff
+
+                val actualInputCrc32 = crc32 & 0xffffffff
+
+                val expectedInputSize =
+                  unsignedToLong(
+                    trailerChunk(4),
+                    trailerChunk(5),
+                    trailerChunk(6),
+                    trailerChunk(7)
+                  )
+
+                if (expectedInputCrc32 != actualInputCrc32) {
+                  F.raiseError[Unit](
+                    new ZipException(
+                      s"Content failed CRC validation: $expectedInputCrc32 != $actualInputCrc32"
+                    )
+                  )
+                } else if (expectedInputSize != actualInputSize) {
+                  F.raiseError[Unit](new ZipException("Content failed size validation"))
+                } else {
+                  F.unit
+                }
+              }
+            }
+          }
+        }
+      }
+      .flatMap(_ => Stream.empty)
+
+  private val gzipHeaderBytes = 10
```

Comment:
I think if we make these ones `final` they can be inlined on JS. See https://github.com/scala-js/scala-js-macrotask-executor/pull/22

Reply:
Made all vals final.